### PR TITLE
Product crud for all product type test cases.

### PR DESCRIPTION
**Pull request by suraj-webkul**
## Issue Reference

Product crud for all product type test cases.

### CODE REVIEW

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 1]
```diff
@@ -40,64 +40,699 @@
...
+        ]);
+    }
 });
 
-it('it should update the existing product', function () {
+it('should return the create page of virtual product', function () {
     // Arrange
     $product = (new ProductFaker())->getSimpleProductFactory()->create();
 
+    $productId = $product->id + 1;
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:225]: Remove this logic.

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 2]
```diff
@@ -40,64 +40,699 @@
...
+        'attribute_family_id' => 1,
+        'sku'                 => $sku = fake()->slug(),
+    ])
+        ->assertOk()
+        ->assertJsonPath('data.redirect_url', route('admin.catalog.products.edit', $productId));
+
+    $this->assertDatabaseHas('products', [
+        'type' => 'virtual',
+        'sku'  => $sku,
+        'id'   => $productId,
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:241]: Remove this assertion.

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 3]
```diff
@@ -30,7 +30,7 @@
         ->assertSeeText(trans('admin::app.catalog.products.index.create-btn'));
 });
 
-it('should return the create page products', function () {
+it('should return the create page of simple product', function () {
     // Arrange
     $product = (new ProductFaker())->getSimpleProductFactory()->create();
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:35]: Remove this logic

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 4]
```diff
@@ -40,64 +40,699 @@
     $this->loginAsAdmin();
 
     postJson(route('admin.catalog.products.store'), [
-        'type'                => $type = fake()->randomElement(['simple', 'configurable', 'virtual', 'downloadable', 'grouped', 'bundle']),
+        'type'                => 'simple',
         'attribute_family_id' => 1,
         'sku'                 => $sku = fake()->uuid(),
     ])
         ->assertOk()
         ->assertJsonPath('data.redirect_url', route('admin.catalog.products.edit', $productId));
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:48]: ->assertJsonPath('data.redirect_url')

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 5]
```diff
@@ -40,64 +40,699 @@
...
 
     $this->assertDatabaseHas('products', [
-        'id'   => $productId,
-        'type' => $type,
+        'type' => 'simple',
         'sku'  => $sku,
+        'id'   => $productId,
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:53]: Remove this.

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 6]
```diff
@@ -40,64 +40,699 @@
...
+        'weight'               => $weight,
+        'locale'               => $locale,
+        'product_id'           => $product->id,
+        'channel'              => $channel,
+    ]);
+});
+
+it('should return the create page of configurable product', function () {
     // Act and Assert
     $this->loginAsAdmin();
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:117]: // Arrange
$attributes = AttributeFamily::find(1)->configurable_attributes

then check this attributes in assertions rathr then hard coding

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 7]
```diff
@@ -40,64 +40,699 @@
...
+    ])
+        ->assertOk()
+        ->assertJsonPath('data.attributes.0.id', 23)
+        ->assertJsonPath('data.attributes.0.code', 'color')
+        ->assertJsonPath('data.attributes.0.options.0.id', 1)
+        ->assertJsonPath('data.attributes.0.options.0.name', 'Red')
+        ->assertJsonPath('data.attributes.1.id', 24)
+        ->assertJsonPath('data.attributes.1.code', 'size')
+        ->assertJsonPath('data.attributes.1.options.0.id', 6)
+        ->assertJsonPath('data.attributes.1.options.0.name', 'S');
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:131]: Check above comment.

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 8]
```diff
@@ -40,64 +40,699 @@
...
+            'price'       => $variant->price,
+            'weight'      => $variant->price,
+            'status'      => $variant->status,
+            'color'       => $variant->color,
+            'size'        => $variant->size,
+            'inventories' => [
+                1 => rand(1111, 9999),
+            ],
+        ];
+    }
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:171]: Remove this logic and make another test case i.e. it should be able to update the variant

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 9]
```diff
@@ -40,64 +40,699 @@
...
+            'short_description'    => $variant->short_description,
+            'description'          => $variant->description,
+            'price'                => $variant->price,
+            'weight'               => $variant->weight,
+            'locale'               => $locale,
+            'channel'              => $channel,
+            'attribute_family_id'  => 1,
+            'visible_individually' => 0,
+        ]);
+    }
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:218]: Remove this and add in new test case

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 10]
```diff
@@ -40,64 +40,699 @@
...
+        'attribute_family_id' => 1,
+        'sku'                 => $sku = fake()->slug(),
+    ])
+        ->assertOk()
+        ->assertJsonPath('data.redirect_url', route('admin.catalog.products.edit', $productId));
+
+    $this->assertDatabaseHas('products', [
+        'type' => 'grouped',
+        'sku'  => $sku,
+        'id'   => $productId,
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:322]: id, type, sku

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 11]
```diff
@@ -40,64 +40,699 @@
...
+        $links['link_' . $key] = [
+            'associated_product_id' => $product->id,
+            'sort_order'            => $key,
+            'qty'                   => rand(10, 100),
+        ];
+    }
+
+    $product = (new ProductFaker())->getSimpleProductFactory()->create([
+        'type' => 'grouped',
+    ]);
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:380]: Setup group product first,

- add the product factory state
- add the group product factory in faker package

it should be like this,

$product = (new ProductFaker())->getGroupedProductFactory()->create();

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 12]
```diff
@@ -40,64 +40,699 @@
...
+        'attribute_family_id' => 1,
+        'sku'                 => $sku = fake()->slug(),
+    ])
+        ->assertOk()
+        ->assertJsonPath('data.redirect_url', route('admin.catalog.products.edit', $productId));
+
+    $this->assertDatabaseHas('products', [
+        'type' => 'downloadable',
         'sku'  => $sku,
+        'id'   => $productId,
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:457]: id, type, sku

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 13]
```diff
@@ -40,64 +40,699 @@
...
+    ]);
+
+    $this->assertDatabaseHas('product_flat', [
+        'url_key'           => $product->url_key,
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:461]: add sku

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 14]
```diff
@@ -40,64 +40,699 @@
...
+        'downloadable_links' => [
+            'link_0' => [
+                'en' => [
+                    'title' => fake()->title,
+                ],
+                'price'            => rand(10, 250),
+                'downloads'        => '1',
+                'sort_order'       => '0',
+                'type'             => 'file',
+                'file'             => $file,
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:534]: $file1

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 15]
```diff
@@ -40,64 +40,699 @@
...
+
+            'link_1' => [
+                'en' => [
+                    'title' => fake()->title,
+                ],
+                'price'            => rand(10, 250),
+                'downloads'        => '1',
+                'sort_order'       => '0',
+                'type'             => 'file',
+                'file'             => $file,
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:548]: $file2

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 16]
```diff
@@ -40,64 +40,699 @@
...
+                'sample_file_name' => $file->getClientOriginalName(),
+            ],
+        ],
+
+        'downloadable_samples' => [
+            'sample_0' => [
+                'title'      => $sample0Title = fake()->title(),
+                'sort_order' => '0',
+                'type'       => 'file',
+                'file'       => $file,
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:561]: $sampleFile1

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 17]
```diff
@@ -40,64 +40,699 @@
...
+                    'sort_order' => $key,
+                    'qty'        => rand(10, 100),
+                ],
+            ],
+        ];
+    }
+
+    $product = (new ProductFaker())->getSimpleProductFactory()->create([
+        'type' => 'bundle',
+    ]);
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:660]: Prepare bundle product factory state.

**Review by devansh-webkul — CHANGES_REQUESTED** [review]
Check the above comments.

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php (hunk 2 above)
Comment (suraj-webkul) [reply packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:241]: done

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (suraj-webkul) [reply packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:53]: done

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php (hunk 6 above)
Comment (suraj-webkul) [reply packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:117]: done

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php (hunk 7 above)
Comment (suraj-webkul) [reply packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:131]: done

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php (hunk 10 above)
Comment (suraj-webkul) [reply packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:322]: done

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php (hunk 11 above)
Comment (suraj-webkul) [reply packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:380]: done

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php (hunk 15 above)
Comment (suraj-webkul) [reply packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:548]: done

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php (hunk 12 above)
Comment (suraj-webkul) [reply packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:457]: done

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php (hunk 13 above)
Comment (suraj-webkul) [reply packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:461]: done

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php (hunk 14 above)
Comment (suraj-webkul) [reply packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php]: done

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php (hunk 17 above)
Comment (suraj-webkul) [reply packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php]: done

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php (hunk 9 above)
Comment (suraj-webkul) [reply packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:259]: done

File: packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php [hunk 18]
```diff
@@ -49,55 +50,718 @@
...
+            'description'          => $variant->description,
+            'price'                => $variant->price,
+            'weight'               => $variant->weight,
+            'locale'               => $locale,
+            'channel'              => $channel,
+            'attribute_family_id'  => 1,
+            'visible_individually' => 0,
+        ]);
+    }
+});
```
Comment (devansh-webkul) [inline packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:260]: Fix this test case as discussed.

Comment (suraj-webkul) [reply packages/Webkul/Admin/tests/Feature/Catalog/ProductTest.php:260]: done

**Review by devansh-webkul — APPROVED** [review]

**Review by jitendra-webkul — APPROVED** [review]

**Pull request merged**